### ISSUE TITLE
Order of loss computation and gradient step seems wrong in vpg.py

### ISSUE THREAD

**Issue by gauravjain14** [openai/spinningup#211]
```
    def update():
        data = buf.get()

        # Get loss and info values before update
        pi_l_old, pi_info_old = compute_loss_pi(data)
        pi_l_old = pi_l_old.item()
        v_l_old = compute_loss_v(data).item()

        # Train policy with a single step of gradient descent
        pi_optimizer.zero_grad()
        loss_pi, pi_info = compute_loss_pi(data)
        loss_pi.backward()
        mpi_avg_grads(ac.pi)    # average grads across MPI processes
        pi_optimizer.step()
```

Here the comment says that train the policy with a single step of gradient descent and then calculate the delta difference.

However, from the code, it seems like that the backward propagation and step() is being performed after both old and the new values of loss are calculated. Shouldn't the following be the order:

```
        # Train policy with a single step of gradient descent
        pi_optimizer.zero_grad()
        pi_l_old.backward()
        mpi_avg_grads(ac.pi)    # average grads across MPI processes
        pi_optimizer.step()
        loss_pi, pi_info = compute_loss_pi(data)
```

Or there is something wrong in my fundamental understanding?

**Comment by jachiam** [comment]
I believe you are right that the calculation of the delta is bugged, and will always yield zero in the current setup. I'll try fixing it over the weekend.

**Comment by gauravjain14** [comment]
Thanks!

**Comment by jachiam** [comment]
No need, I'll get it figured out.
